src/svm/coders.ts

```diff
@@ -0,0 +1,267 @@
+      getOrInsertDefault(ix.programId);
+      for (const accountMeta of ix.keys) {
+        const keyMode = getOrInsertDefault(accountMeta.pubkey);
+        keyMode.isWritable ||= accountMeta.isWritable;
+      }
+    }
+
+    return new UnsignedCompiledKeys(keyModeMap, payer);
+  }
+
+  getMessageComponents(): [MessageHeader, PublicKey[]] {
+    const mapEntries = [...this.keyModeMap.entries()];
+    if (mapEntries.length > 256) throw new Error("Max static account keys length exceeded");
+
+    const writableNonSigners = mapEntries.filter(([, mode]) => mode.isWritable);
+    const readonlyNonSigners = mapEntries.filter(([, mode]) => !mode.isWritable);
+
+    const header: MessageHeader = {
+      numRequiredSignatures: 0,
+      numReadonlySignedAccounts: 0,
+      numReadonlyUnsignedAccounts: readonlyNonSigners.length,
+    };
+
+    const staticAccountKeys = [
+      ...writableNonSigners.map(([address]) => new PublicKey(address)),
+      ...readonlyNonSigners.map(([address]) => new PublicKey(address)),
+    ];
+
+    return [header, staticAccountKeys];
+  }
+}
+
+/**
+ * Extended version of legacy Message to handle compilation of unsigned transactions. Base implementation is here:
+ * https://github.com/solana-labs/solana-web3.js/blob/v1.95.3/src/message/legacy.ts
+ */
+class UnsignedMessage extends Message {
+  static compileUnsigned(instructions: TransactionInstruction[], payer?: PublicKey): Message {
+    const compiledKeys = UnsignedCompiledKeys.compileUnsigned(instructions, payer);
+    const [header, staticAccountKeys] = compiledKeys.getMessageComponents();
+    const accountKeys = new MessageAccountKeys(staticAccountKeys);
+    const compiledInstructions = accountKeys.compileInstructions(instructions).map(
+      (ix: MessageCompiledInstruction): CompiledInstruction => ({
+        programIdIndex: ix.programIdIndex,
+        accounts: ix.accountKeyIndexes,
+        data: bs58.encode(ix.data),
+      })
+    );
+    return new Message({
+      header,
+      accountKeys: staticAccountKeys,
+      recentBlockhash: "", // Not used as we are not signing the transaction.
+      instructions: compiledInstructions,
+    });
+  }
+}
+
+/**
+ * Helper to encode MulticallHandler transactions.
+ */
+export class MulticallHandlerCoder {
+  readonly compiledMessage: Message;
+
+  private readonly layout: Layout;
+
+  constructor(instructions: TransactionInstruction[], payerKey?: PublicKey) {
+    // Compile transaction message and keys.
+    this.compiledMessage = UnsignedMessage.compileUnsigned(instructions, payerKey);
+
+    // Setup the layout for the encoder.
+    const fieldLayouts = [IdlCoder.fieldLayout(MulticallHandlerCoder.coderArg, MulticallHandlerCoder.coderTypes)];
+    this.layout = borsh.struct(fieldLayouts);
+  }
+
+  private static coderArg = {
+    name: "compiledIxs",
+    type: {
+      vec: {
+        defined: {
+          name: "compiledIx",
+        },
+      },
+    },
+  };
+
+  private static coderTypes: IdlTypeDef[] = [
```

Comment:
this can all be linted a lot better. might be ok to ignore though and we can run a seperate linting task to collaps JS objects after the fact.